### PR TITLE
GC: factor out visiting all machine values

### PR DESCRIPTION
**Pull request by RalfJung**
@saethlin that is roughly what I had in mind.

I think some parts of the state are skipped by the visitor. I listed the ones that I found in FIXMEs but I am not sure if that list is complete.

### CODE REVIEW

**Comment by RalfJung** [comment]
More bits of machine state that are not visited:
- last_error
- catch_unwind

**Comment by RalfJung** [comment]
We will probably want a `VisitMachineValues` trait that is implemented for `Machine` and recursively for all the non-trivial types it has as fields. That'll be quite a bit of work, but the current approach clearly misses too many bits of the state.
```rust
trait VisitMachineValues {
  fn visit_machine_values(&self, impl FnMut(&Operand<Provenance>));
}
```

**Comment by saethlin** [comment]
I can read what you changed, but I don't understand how this helps with missing pieces of state in the interpreter?

**Comment by RalfJung** [comment]
I think separating the visiting from the remaining GC logic makes it easier to check that nothing was forgotten. Also using the visitor pattern with exhaustive field lists in the `let` bindings (see my latest commit) means we get errors when a field was skipped.

**Comment by RalfJung** [comment]
I am happy with the structure of this PR now. @saethlin do you want to take over and complete the PR by implementing all the missing things in the visitor? Or should we land this and you hopefully can do that in a future PR?

File: src/concurrency/thread.rs [hunk 1]
```diff
@@ -628,6 +624,34 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
...
+            for frame in &thread.stack {
+                // Return place.
+                if let Place::Ptr(mplace) = *frame.return_place {
+                    visit(&Operand::Indirect(mplace));
+                }
+                // Locals.
+                for local in frame.locals.iter() {
+                    let LocalValue::Live(value) = &local.value else {
+                    continue;
+                };
```
Comment (saethlin) [inline src/concurrency/thread.rs:647]: This identation threw me for a loop

File: src/concurrency/thread.rs [hunk 2]
```diff
@@ -628,6 +624,34 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
     }
 }
 
+impl VisitMachineValues for ThreadManager<'_, '_> {
+    fn visit_machine_values(&self, visit: &mut impl FnMut(&Operand<Provenance>)) {
+        // FIXME some other fields also contain machine values
```
Comment (saethlin) [inline src/concurrency/thread.rs:629]: I see there are other fields with machine values, but it's not clear to me that for example `thread_local_alloc_ids` is keeping pointers live. Are there other fields?

**Comment by saethlin** [comment]
I'd prefer to do a second PR. r=me with or without addressing the formatting nit (I think we have other formatting jank which will snap into place when style team gets off the ground).

File: src/concurrency/thread.rs (hunk 1 above)
Comment (RalfJung) [reply src/concurrency/thread.rs:647]: Oh I didn't even realize... good catch.

File: src/concurrency/thread.rs (hunk 2 above)
Comment (RalfJung) [reply src/concurrency/thread.rs:629]: Isn't that the only way you will reach the current value of thread-local variables? And a thread-local variable can store a pointer to somewhere else.

Comment (RalfJung) [reply src/concurrency/thread.rs:629]: It's similar, I think, to the global base tag map in Stacked Borrows -- I don't know how your GC is not failing left and right due to removing the base tags.

Comment (saethlin) [reply src/concurrency/thread.rs:629]: `Stack::retain` never removes the bottom-most tag.

Comment (RalfJung) [reply src/concurrency/thread.rs:629]: Ah. Yeah that would help but needs to be carefully documented.

**Pull request merged**